src/main/java/org/elasticsearch/plugin/zentity/ModelsAction.java

```diff
@@ -600,22 +645,234 @@ public void onFailure(Exception e) {
+                        },
+                        errorHandler
+                    ));
                 }
 
             } catch (Exception e) {
-                ZentityPlugin.sendResponseError(channel, logger, e);
+                errorHandler.accept(e);
             }
         };
     }
-}
+
+    /**
+     * Create a single result containing the error.
+     *
+     * @param delegate The delegated action listener to run after creating the single error result.
+     * @param action   The bulk action ("create", "update", "delete").
+     * @param e        The exception object.
+     */
+    static void delegateFailure(ActionListener<BulkAction.SingleResult> delegate, String action, Exception e) {
+        try {
+            delegate.onResponse(new BulkAction.SingleResult("{\"" + action + "\":{\"error\":{" + Job.serializeException(e, true) + "}}}", true));
+        } catch (Exception ee) {
+            // An error occurred when preparing or sending the response.
+            delegate.onFailure(ee);
+        }
+    }
+
+    /**
+     * Run a collection of operations concurrently.
+     *
+     * @param client The node client.
+     * @param entries The bulk tuple entries: <String actionAndParams, String entityModel>.
+     * @param reqParams The parameters map for the entire request. Overridden by any params from entries.
+     * @param listener The listener for completion results.
+     */
+    static void executeBulk(NodeClient client, List<Tuple<String, String>> entries, Map<String, String> reqParams, ActionListener<Collection<BulkAction.SingleResult>> listener) {
+
+        // Process a single bulk entry.
+        BiConsumer<Tuple<String, String>, ActionListener<BulkAction.SingleResult>> operationRunner = (tuple, delegate) -> {
+            String actionAndParams = tuple.v1();
+            String entityModel = tuple.v2();
+            String action = "action";
+            String params = "";
+            try {
+                Iterator<Map.Entry<String, JsonNode>> fields = Json.MAPPER.readTree(actionAndParams).fields();
+                while (fields.hasNext()) {
+                    Map.Entry<String, JsonNode> field = fields.next();
+                    String name = field.getKey();
+                    JsonNode value = field.getValue();
+                    switch (name) {
+                        case "create":
+                        case "update":
+                        case "delete":
+                            if (!action.equals("action"))
+                                throw new ValidationException("Each bulk operation must have only one action and payload.");
+                            action = name;
+                            params = Json.ORDERED_MAPPER.writeValueAsString(value);
+                            break;
+                        default:
+                            throw new ValidationException("'" + name + "' is not a recognized action for bulk model management.");
```

Comment:
nit, might be nicer to send the error back to the listener here explicitly instead of throwing and then catching below